src/lib/modules/scaffolding/contractLanguage/solidity/index.ts

```diff
@@ -0,0 +1,73 @@
+import Handlebars from "handlebars";
+import * as path from "path";
+
+import { Embark } from "../../../../../typings/embark";
+import { Builder } from "../../builder";
+import { Description } from "../../description";
+import { Options } from "../../options";
+
+const fs = require("../../../../core/fs");
+require("../../handlebarHelpers");
+
+const templatePath = path.join(__dirname, "templates", "contract.sol.hbs");
+
+export class Solidity implements Builder {
+  constructor(private embark: Embark,
+              private description: Description,
+              private options: Options) {
+  }
+
+  public readonly build = async () => {
+    return Object.keys(this.description.data).map((contractName) => {
+      const code = this.generateCode(contractName);
+      const file = this.saveFile(contractName, code);
+      this.printInstructions(contractName);
+      return file;
+    });
+  }
+
+  private readonly generateCode = (contractName: string) => {
+    const source = fs.readFileSync(templatePath, "utf-8");
+    const template = Handlebars.compile(source);
+
+    const attributes = this.description.standardAttributes(contractName);
+    const ipfs = this.description.ipfsAttributes(contractName);
+    const associations = this.description.associationAttributes(contractName);
+
+    const data = {
+      associations,
+      attributes,
+      contractName,
+      ipfs,
+      structName: `${contractName}Struct`,
+    };
+
+    return template(data);
+  }
+
+  private readonly saveFile = (contractName: string, code: string) => {
+    const filename = `${contractName}.sol`;
+    const contractDirs = this.embark.config.embarkConfig.contracts;
+    const contractDir = Array.isArray(contractDirs) ? contractDirs[0] : contractDirs;
+    const filePath = fs.dappPath(contractDir.replace(/\*/g, ""), filename);
+    if (!this.options.overwrite && fs.existsSync(filePath)) {
+      this.embark.logger.error(__(`The contract ${contractName} already exists, skipping.`));
+      return;
+    }
+
+    fs.writeFileSync(filePath, code);
+    return filePath;
+  }
+
+  private readonly printInstructions = (contractName: string) => {
+    const associations = Object.keys(this.description.associationAttributes(contractName));
+    if (!associations.length) {
+      return;
+    }
+
+    const args = associations.map((name) => `"$${name}"`).join(", ");
+    this.embark.logger.info(`In order to deploy your contracts, you will have to specify the dependencies.`);
+    this.embark.logger.info(`You can do it by adding to your contracts config the following snippets:`);
+    this.embark.logger.info(`${contractName}: { args: [${args}] }`);
+  }
+}
```

Comment:
This looks very solid! One question though: I've noticed you're doing property initialization instead of method declaration. E.g.

```
public readonly build = async () => {

}
```

vs

```
async build() {

}
```
This has a couple of side effects including that the method is not generated on the prototype of the resulting instance, but on every single instance that is created (that's what we usually want to avoid unless you have reason to do otherwise).

I was also wondering why we're using `readonly` here, considering that those "properties" are actually all functions anyways.

Reply:
my bad, I didn't know, I have updated